be/src/agent/task_worker_pool.cpp

```diff
@@ -618,6 +619,116 @@ void* TaskWorkerPool::_push_worker_thread_callback(void* arg_this) {
 
 void* TaskWorkerPool::_publish_version_worker_thread_callback(void* arg_this) {
     TaskWorkerPool* worker_pool_this = (TaskWorkerPool*)arg_this;
+
+    std::unique_ptr<ThreadPool> threadpool;
+    auto st = ThreadPoolBuilder("publish_version")
+                      .set_min_threads(config::partition_publish_version_worker_count)
+                      .set_max_threads(config::partition_publish_version_worker_count)
+                      .set_max_queue_size(32)
+                      .set_idle_timeout(MonoDelta::FromMilliseconds(2000))
+                      .build(&threadpool);
+
+    // Use a lambda to make code clearer.
+    auto publish_version_in_parallel = [&worker_pool_this, &threadpool](
+                                               const TPublishVersionRequest& publish_version_req,
+                                               std::vector<TTabletId>* error_tablet_ids) -> Status {
+        int64_t transaction_id = publish_version_req.transaction_id;
+        Status error_status = Status::OK();
+
+        // each partition
+        for (auto& par_ver_info : publish_version_req.partition_version_infos) {
+            int64_t partition_id = par_ver_info.partition_id;
+            // get all partition related tablets and check whether the tablet have the related version
+
+            std::set<TabletInfo> partition_related_tablet_infos;
+            if (publish_version_req.strict_mode) {
+                StorageEngine::instance()->tablet_manager()->get_partition_related_tablets(
+                        partition_id, &partition_related_tablet_infos);
+                if (partition_related_tablet_infos.empty()) {
+                    VLOG(1) << "could not find related tablet for partition " << partition_id
+                            << ", skip publish version";
+                    continue;
+                }
+            }
+
+            map<TabletInfo, RowsetSharedPtr> tablet_related_rs;
+            StorageEngine::instance()->txn_manager()->get_txn_related_tablets(transaction_id, partition_id,
+                                                                              &tablet_related_rs);
+
+            TVersion version = par_ver_info.version;
+
+            std::vector<TabletInfo> tablet_infos;
+            tablet_infos.reserve(tablet_related_rs.size());
+            std::vector<Status> statuses(tablet_related_rs.size(), Status::OK());
+
+            size_t idx = 0;
+            // each tablet
+            for (auto& tablet_rs : tablet_related_rs) {
+                tablet_infos.push_back(tablet_rs.first);
+
+                threadpool->submit_func([&worker_pool_this, &tablet_rs, &statuses, idx, &version, &transaction_id,
```

Comment:
submit_func may return ServiceUnavailable, handle it

Reply:
Updated. status is checked, and would retry task after a while.